[WIP][#607] Edit an existing migration plan

mturley opened the pull request:
WIP PR for starting on #607 

- [x] Add kebab menu with edit option to plans not started list
- [ ] Figure out how to open the plan wizard in edit mode for a specific plan
- [ ] Open the plan wizard in edit mode
- [ ] When the plan wizard is in edit mode, pre-fill selections
- [ ] ???
- [ ] When the plan wizard is submitted in edit mode, use edit API endpoint instead of creating a new plan

mturley commented:
As soon as @AparnaKarve's PR https://github.com/ManageIQ/manageiq/pull/17976 is merged, I'll be able to fix the path column in the VM step, and then this PR will be ready for review.

michaelkro commented on app/javascript/react/screens/App/Overview/screens/PlanWizard/PlanWizard.js:
```diff
@@ -125,6 +126,18 @@ class PlanWizard extends React.Component {
       hideAlertAction();
+
+      if (editingPlan && planWizardGeneralStep.values.vm_choice_radio === 'vms_via_csv') {
+        // If the user selected CSV and they are editing a plan, warn before proceeding.
```
Please remove comment

michaelkro commented on app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardAdvancedOptionsStep/PlanWizardAdvancedOptionsStep.js:
```diff
@@ -99,13 +105,6 @@ PlanWizardAdvancedOptionsStep.defaultProps = {
-    postMigrationPlaybook: ''
-  },
   destroyOnUnmount: false
+  // initialValues prop is passed in via ./index
```
I think this guy can be removed too

michaelkro commented:
Sort of an weird interaction between CSV import auto selecting all valid VMs and edit pre-fill.  Not sure if this is a non-issue though @vconzola?

# Steps to Reproduce
1.  Create a plan using VM discovery, select a few VMs to migrate
2.  Edit the plan, but this time use CSV import, making sure that at least one of the previously selected VMs is in your CSV
3.  On the VMs step, the previously selected VM, i.e., the VM that should be pre-selected for edit will not be selected.  This is b/c it is considered to already be part of a plan and is in the "warn but don't block state"

vconzola commented:
@michaelkro Not sure what you mean by in the "warn but don't block state" for VMs. Maybe you can show me? But this has me re-thinking about how we don't allow VMs to be in more than one plan. Is this really what we want? Also, what happens if we delete a plan before it's run - do the unmigrated VMs get released back into the wild?

michaelkro commented:
@vconzola here is a screen

<img width="1920" alt="fullscreen_9_13_18__9_08_am" src="https://user-images.githubusercontent.com/15141412/45490372-e4d2c200-b734-11e8-9ad6-d8e67c67cf29.png">

I'm still kind of confused as to which VMs can and cannot be discovered, to be honest.  The results seem to vary for me depending on the status of the plan in which the VM is included, @AparnaKarve can you bring us into the light with regard to this?

I do believe that if a plan is deleted before running, its VMs become re-discoverable though

michaelkro commented on app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardGeneralStep/PlanWizardGeneralStep.js:
```diff
@@ -8,7 +8,8 @@ import { BootstrapSelect } from '../../../../../common/forms/BootstrapSelect';
 import { asyncValidate, onChange } from './helpers';
 
-const PlanWizardGeneralStep = ({ transformationMappings }) => (
+// TODO disable the mapping selection when editing
```
I think this `TODO` is `TODONE`

mturley replied:
Oops! Thanks.

AparnaKarve commented:
> The results seem to vary for me depending on the status of the plan in which the VM is included, @AparnaKarve can you bring us into the light with regard to this?

@michaelkro Sure. I can get more info on this based on how the backend code is written and my own observations.

AparnaKarve commented:
@mturley Maybe we should not allow editing a plan that does not have an associated infrastructure mapping - because we get this error during the VM discovery step.

<img width="908" alt="screen shot 2018-09-13 at 8 24 03 am" src="https://user-images.githubusercontent.com/1538216/45498797-c4543900-b72f-11e8-9fc4-d2be4395f88d.png">

AparnaKarve commented:
The pre-selected VM in the VM Step shows up twice in Advanced Options

<img width="896" alt="screen shot 2018-09-13 at 8 37 12 am" src="https://user-images.githubusercontent.com/1538216/45499053-63793080-b730-11e8-8225-c2f0042bbe43.png">


<img width="905" alt="screen shot 2018-09-13 at 8 36 56 am" src="https://user-images.githubusercontent.com/1538216/45499037-58be9b80-b730-11e8-8ce9-1543cec1bd2a.png">

mturley commented:
@AparnaKarve I'm confused, how did you get a plan that has no associated mapping? I thought that was required when creating a plan?

mturley commented:
Thanks for reporting the other problem; it looks like this only happens when editing a second time, and only for the VMs that were added when editing the first time. I wasn't able to reproduce it with the first edit. I'll look into it

AparnaKarve commented:
>  I'm confused, how did you get a plan that has no associated mapping? I thought that was required when creating a plan?

@mturley I picked one of the aug21 plan from this list -

<img width="1181" alt="screen shot 2018-09-13 at 8 44 28 am" src="https://user-images.githubusercontent.com/1538216/45499423-42fda600-b731-11e8-8755-0bad32360b08.png">

michaelkro commented:
Advanced Options Step Table selected count can get confused during edit

# Screens
<img width="1920" alt="fullscreen_9_13_18__11_42_am" src="https://user-images.githubusercontent.com/15141412/45499429-6bde6500-b74a-11e8-8332-b33821eb1d97.png">

# Steps to Reproduce
- Create a plan configured to run a pre or post playbook on some VMs
- Edit the plan, and on the VMs step, remove all the pre-selected VMs and choose some different ones
- Notice that the count at the bottom of the select column seems to be reflecting pre-loaded values from the currently being edited plan

AparnaKarve commented:
> Maybe we should not allow editing a plan that does not have an associated infrastructure mapping - because we get this error during the VM discovery step.

Or the other option for this case would be to enable the Infra Mapping dropdown, and allow the user to select one.

michaelkro commented:
VMs reverting to preselected values when moving backwards from Advanced Options Step

# Steps to Reproduce
- Open an existing plan in edit mode
- On the VMs step, edit the selected VMs and navigate forward to the Advanced Options Step
- Navigate back to the VMs step and notice that the table has reverted to the pre-selected values

AparnaKarve commented:
> The results seem to vary for me depending on the status of the plan in which the VM is included, @AparnaKarve can you bring us into the light with regard to this?

| Use case  | VM visible in discovery mode| VM visible in CSV import mode |
| ------------- | ------------- | ------------- |
| VM not included in any plan  | Yes | Yes |
| VM included in some plan  | No  | Yes, with a warning that VM exists in a different plan |
| Plan with VM is deleted  | Yes, in subsequent VM discoveries after plan deletion  | Yes  |
| Plan with VM fails migration  | Yes, in subsequent VM discoveries after the plan fails  | Yes  |
| Plan with VM completes migration successfully | No  | Yes, with a warning that VM already migrated  |

My test results do match the above table

mturley commented:
@michaelkro, I opened https://github.com/ManageIQ/manageiq-v2v/issues/639 to address https://github.com/ManageIQ/manageiq-v2v/pull/620#issuecomment-421058720, and I addressed https://github.com/ManageIQ/manageiq-v2v/pull/620#issuecomment-421055126 in https://github.com/ManageIQ/manageiq-v2v/pull/620/commits/f0f10bc102d7b4ee0cef52bfb909681c77a4808f

@AparnaKarve, I addressed https://github.com/ManageIQ/manageiq-v2v/pull/620#issuecomment-421052404 in https://github.com/ManageIQ/manageiq-v2v/pull/620/commits/81cb59a01b4b6a1397fd6d0fe8ec2b4d53462913, and https://github.com/ManageIQ/manageiq-v2v/pull/620#issuecomment-421050631 in https://github.com/ManageIQ/manageiq-v2v/pull/620/commits/616747466e3a390b66921a2717a32af53528f8f4.

I think the only remaining feedback I have not addressed here is https://github.com/ManageIQ/manageiq-v2v/pull/620#issuecomment-420993932, and I'm not sure whether we need to do anything about that here, @vconzola ?

mturley commented:
@michaelkro, I opened https://github.com/ManageIQ/manageiq-v2v/issues/640 to track the remaining issue so we can merge this this sprint.

michaelkro commented on app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardResultsStep/PlanWizardResultsStepActions.js:
```diff
@@ -35,3 +39,21 @@ const _postMigrationPlansActionCreator = (url, migrationPlans, planSchedule) =>
+    })
+  });
+
+export const putMigrationPlansAction = (url, migrationPlans, planSchedule) =>
```
There is something going on with this function signature.  Should there be four arguments with the second being `id`?

mturley replied:
Wow, how did that even work... I guess it was dropping the actual `planSchedule` argument, thanks for that catch!

AparnaKarve commented:
Noticed that the wizard opens up when I click on the disabled Edit menu item.

AparnaKarve commented:
There seems to be a discrepancy in the VM discovery process with Create and Edit (and this is probably a backend issue)

The table above https://github.com/ManageIQ/manageiq-v2v/pull/620#issuecomment-421084070 shows the behavior during a Create Plan operation.
I would have expected the same behavior for Edit, but that does not seem to be the case.

Steps to reproduce
1. Create a plan with VM `A`
2. Edit the plan and add VM `B` and save the plan.  This plan now has VMs `A` and `B`
3. Create another plan. The discovery process in this Create operaion seems to be discovering `B`

michaelkro commented on app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardAdvancedOptionsStep/index.js:
```diff
@@ -2,15 +2,17 @@ import { connect } from 'react-redux';
-import { getVMStepSelectedVms } from './PlanWizardAdvancedOptionsStepSelectors';
+import { getVMStepSelectedVms, getVmIdsWithProperty } from './helpers';
 import reducer from './PlanWizardAdvancedOptionsStepReducer';
+import { findEditingPlan } from '../../PlanWizardSelectors';
```
Since we use this helper to add `editingPlan` to `mapStateToProps` in multiple containers, might it make sense to do this only once?

Maybe in `PlanWizard/index.js` or `Overview/index.js`.  Then we could just read the `editingPlan` from the redux store in the steps that need it

mturley replied:
That's a good point. I'll do that.

mturley replied:
(Note: after discussion with @michaelkro we decided to leave this as-is)

mturley commented:
> Create a plan with VM A
> Edit the plan and add VM B and save the plan. This plan now has VMs A and B
> Create another plan. The discovery process in this Create operaion seems to be discovering B

@AparnaKarve I noticed that too, and in fact it was discovering B again even when editing the plan, which is why I filtered out duplicates here: https://github.com/ManageIQ/manageiq-v2v/pull/620/files#diff-5c1096a11ccff6f726753cbd48bc02f8R130

I'm not sure what's going on there, I also suspect it's a backend issue.

michaelkro commented:
Seeing a possible error attempting to actually run an edited plan

<img width="1920" alt="fullscreen_9_14_18__10_39_am" src="https://user-images.githubusercontent.com/15141412/45556989-95f75c00-b80a-11e8-841b-4836df5270c8.png">

@mturley @AparnaKarve can you confirm whether or not you can migrate?

AparnaKarve commented:
Interesting test @michaelkro !

I think the Edit `service_template` API needs to be looked at...
It could very well be responsible for all the VM discovery issues we are seeing and now even the actual plan execution.

mturley commented:
@AparnaKarve @michaelkro I can confirm these issues with running an edited plan. I tried converting the PUT to a POST with action: edit, and I'm still getting the same issues. I'll push that commit anyway, since that seems like the pattern for editing things in this API.

AparnaKarve commented:
So this is indeed a backend bug.

When you save a Plan with Create or Edit it is supposed to create a `ServiceResource` record for the VM.
Currently, that is happening only for Create, but not for Edit.

AparnaKarve commented:
Backend PR to fix the VM discovery issues - https://github.com/ManageIQ/manageiq/pull/17989
